### PR TITLE
Add GameMode shared module

### PR DESCRIPTION
**Pull request by orowith2os**
Particularly useful for a number of apps on Flathub, such as Lutris, Steam, Bottles, and WarThunder.

I've separated the libraries and script so that apps can add the bare minimum of what they need.

### CODE REVIEW

File: gamemode/gamemoderun.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+  "name": "gamemoderun",
+  "buildsystem": "simple",
+  "build-commands": [
+    "mv data/gamemoderun /app/bin/gamemoderun",
```
Comment (TingPing) [inline gamemode/gamemoderun.json:5]: This should be built from source.

Comment (TingPing) [reply gamemode/gamemoderun.json:5]: Oh its just a shell script.

You could simplify this by only downloading a single file.

Comment (orowith2os) [reply gamemode/gamemoderun.json:5]: It doesn't need to be, as it's a shell script. I tried using the `install` command for this, but couldn't quite figure it out, so I fell back to `mv` and `chmod`.

File: gamemode/README.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+This is a submodule for FeralInteractive's GameMode.
+
+To use this, add the manifests listed here; `libgamemode*.json` will install the shared libraries, `gamemoderun.json` will install `gamemoderun` to `/app/bin`.
+
+You will need to add both manifests separately.
```
Comment (TingPing) [inline gamemode/README.md:5]: What's the harm in `libgamemode` always installing `gamemoderun`?

Comment (orowith2os) [reply gamemode/README.md:5]: with libgamemode, apparently the meson buildsystem doesn't install gamemoderun to the prefix, no matter the config-opts. I can make an issue in upstream gamemode about it, or see if I can find something about why it's not done.

File: gamemode/libgamemode-1.7.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    "-Dwith-sd-bus-provider=no-daemon" 
+  ],
+  "sources": [
+    {
+      "type": "archive",
+      "url": "https://github.com/FeralInteractive/gamemode/archive/refs/tags/1.7.tar.gz",
+      "sha256": "8a1718c657cb1d3c8269edfd397faf3cb41fccd885198ef0225bf8f82217c42f"
+    }
+  ],
+  "cleanup": [ "/include", "/lib/pkgconfig", "/lib/libgamemodeauto.a" ]
```
Comment (TingPing) [inline gamemode/libgamemode-1.7.json:16]: Try `-Ddefault_library=shared` and see if it still builds a static lib.

File: gamemode/README.md (hunk 2 above)
Comment (orowith2os) [reply gamemode/README.md:5]: I'll merge the two manifests, it's not a large deal to add a separate module for gamemoderun anyways.

File: gamemode/libgamemode-1.7.json (hunk 3 above)
Comment (TingPing) [reply gamemode/libgamemode-1.7.json:16]: Oh I see it hardcodes builting both libraries.

Change this to a `"post-install"` and `rm` the shared library. We don't want projects accidentally linking to it.

File: gamemode/README.md (hunk 2 above)
Comment (TingPing) [reply gamemode/README.md:5]: You can add a custom `"post-install"` command to install the script also.

Comment (TingPing) [reply gamemode/README.md:5]: Looking at their meson files they correctly install it into the right prefix..

Comment (TingPing) [reply gamemode/README.md:5]: It was fixed in https://github.com/FeralInteractive/gamemode/commit/85d75c303c61d75602a8415716aa9d4befbf6f8a

Comment (orowith2os) [reply gamemode/README.md:5]: > Looking at their meson files they correctly install it into the right prefix..

Interesting, it didn't do that when I built it without the config-opts. Is there anything I need to do to get it to install?

Comment (TingPing) [reply gamemode/README.md:5]: It is behind the `if sd_bus_provider != 'no-daemon'` option.

Comment (TingPing) [reply gamemode/README.md:5]: I think their logic is just wrong. A manual install seems fine.

**Comment by orowith2os** [comment]
Simplified it a bit, so now it handles everything in one manifest.

File: gamemode/gamemode-1.7.json [hunk 4]
```diff
@@ -13,5 +13,11 @@
       "sha256": "8a1718c657cb1d3c8269edfd397faf3cb41fccd885198ef0225bf8f82217c42f"
     }
   ],
-  "cleanup": [ "/include", "/lib/pkgconfig", "/lib/libgamemodeauto.a" ]
+  "cleanup": [ "/include", "/lib/pkgconfig" ],
+  "post-install": [
+    "mkdir -p /app/bin/",
+    "mv ../data/gamemoderun /app/bin/gamemoderun",
+    "chmod +x /app/bin/gamemoderun",
```
Comment (TingPing) [inline gamemode/gamemode-1.7.json:20]:
```suggestion
    "install -t /app/bin -Dm755 ../data/gamemoderun",
```

This probably works.

File: gamemode/gamemode-1.7.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+      "type": "archive",
+      "url": "https://github.com/FeralInteractive/gamemode/archive/refs/tags/1.7.tar.gz",
+      "sha256": "8a1718c657cb1d3c8269edfd397faf3cb41fccd885198ef0225bf8f82217c42f"
+    }
+  ],
+  "cleanup": [ "/include", "/lib/pkgconfig" ],
+  "post-install": [
+    "mkdir -p /app/bin/",
+    "mv ../data/gamemoderun /app/bin/gamemoderun",
+    "chmod +x /app/bin/gamemoderun",
```
Comment (TingPing) [inline gamemode/gamemode-1.7.json:20]:
```suggestion
    "install -t /app/bin -Dm755 ../data/gamemoderun",
```

This probably works.

**Comment by A6GibKm** [comment]
By the way, for lutris we also do a 32bit build (is this needed like at all?), this would make this shared module useless for us.

**Comment by orowith2os** [comment]
> By the way, for lutris we also do a 32bit build (is this needed like at all?), can that be done with this shared-module?

I don't *believe* a 32-bit build is needed, but I could look into doing a 32-bit one as well. Though, it might be a bit weird, as apparently Lutris adds an ld flag and libdir.... TingPing might be able to give me some tips on that, if those are needed.

<https://github.com/flathub/net.lutris.Lutris/blob/beta/net.lutris.Lutris.yml#L70>
<https://github.com/flathub/net.lutris.Lutris/blob/beta/net.lutris.Lutris.yml#L283>

**Comment by orowith2os** [comment]
Could I have some tips? I don't know how I would go about adding the 32-bit variant, especially since it seems to need some modifications to the ldflags and whatnot.

**Comment by Doomsdayrs** [comment]
@orowith2os I sent a PR to your repo, Let me get to work

**Comment by Doomsdayrs** [comment]
> Could I have some tips? I don't know how I would go about adding the 32-bit variant, especially since it seems to need some modifications to the ldflags and whatnot.

Lutris seems to do it via some flags, which seems very application specific

**Comment by orowith2os** [comment]
Update comment: It seems like I can add 32-bit builds fairly easily, but it might be safer to separate them into their own manifest? 

I'll also stop versioning this when I get to updating it, for the same reasons I'll stop versioning the SDL2 module: there isn't much reason to it, since it's more or less compatible across versions.

**Comment by RobLoach** [comment]
Was looking for this one. Any updates needed?

File: gamemode/gamemode-1.7.json [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    {
+      "type": "archive",
+      "url": "https://github.com/FeralInteractive/gamemode/archive/refs/tags/1.7.tar.gz",
+      "sha256": "8a1718c657cb1d3c8269edfd397faf3cb41fccd885198ef0225bf8f82217c42f"
+    }
+  ],
+  "cleanup": [ "/include", "/lib/pkgconfig" ],
+  "post-install": [
+    "install -t /app/bin -Dm755 ../data/gamemoderun",
+    "rm /app/lib/libgamemodeauto.a"
```
Comment (ranisalt) [inline gamemode/gamemode-1.7.json:19]: Add `/lib/*.a` to `cleanup` instead

**Comment by orowith2os** [comment]
> Any updates needed?

I need to manage 32-bit builds somehow, and I'm not sure on how to do that. I also need to fix up the cleanup stuff :P

I'll get to this when I have a computer set up

File: gamemode/gamemode-1.7.json [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    "-Dwith-sd-bus-provider=no-daemon"
+  ],
+  "sources": [
+    {
+      "type": "archive",
+      "url": "https://github.com/FeralInteractive/gamemode/archive/refs/tags/1.7.tar.gz",
+      "sha256": "8a1718c657cb1d3c8269edfd397faf3cb41fccd885198ef0225bf8f82217c42f"
+    }
+  ],
+  "cleanup": [ "/include", "/lib/pkgconfig" ],
```
Comment (RobLoach) [inline gamemode/gamemode-1.7.json:16]:
``` suggestion
  "cleanup": [ 
    "/include",
    "/lib/pkgconfig",
    "/lib/*.a"
  ],
```

File: gamemode/gamemode-1.7.json (hunk 6 above)
Comment (RobLoach) [inline gamemode/gamemode-1.7.json:19]:
``` suggestion
    "install -t /app/bin -Dm755 ../data/gamemoderun"
```

Comment (hfiguiere) [reply gamemode/gamemode-1.7.json:19]: Just add `*.a` like everywhere else.

**Comment by orowith2os** [comment]
I'm still not sure how I should handle 32-bit builds, until that's figured out, this is blocked.

**Comment by Doomsdayrs** [comment]
> I'm still not sure how I should handle 32-bit builds, until that's figured out, this is blocked.

Who can benefit from the 32bit build currently?

**Comment by orowith2os** [comment]
> Who can benefit from the 32bit build currently?

Basically anybody Wine-based.

**Comment by orowith2os** [comment]
I'll close this in favor of https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/issues/1641.

**Pull request closed without merging**